src/build.rs

```diff
@@ -130,105 +134,146 @@ impl BuildQueue {
+    // obsolete, and if there are more recent or higher priority requests waiting.
+    // To make these checks we compare timestamps. We have the timestamp for the
+    // current request and the most recently completed request, so checking
+    // obsolete-ness is straightforward.
+    //
+    // Before we wait for the lock, the requesting thread registers that it is waiting
+    // by recording its request time in `queued_start_times`. When a thread aquires
+    // the lock, it checks if a more worthy thread is queued and if so it squashes
+    // itself.
+    //
+    // After building, we must carefully set the appropriate timestamps.
+    pub fn request_build(&self, new_build_dir: &Path, mut priority: BuildPriority) -> BuildResult {
+        let start_time = Instant::now();
+        trace!("request_build, {:?} {:?}", new_build_dir, priority);
+
+        let build_dir = {
+            let mut build_dir = self.build_dir.lock().unwrap();
+            // If the build dir changes, upgrade the request to high priority.
+            if build_dir.as_ref().map_or(true, |dir| dir != new_build_dir) {
+                priority = BuildPriority::Cargo;
+                *build_dir = Some(new_build_dir.to_owned());
             }
-        }
-
-        self.cancel_pending();
+            // Keep a copy of the build directory in case the field is stomped by another thread.
+            build_dir.as_ref().unwrap().clone()
+        };
 
-        let (tx, rx) = channel();
-        self.pending.lock().unwrap().push(tx);
+        // Record our request time.
         if priority == BuildPriority::Normal {
+            // Normal priority threads sleep before starting up.
+            // Note that we sleep *before* setting a pending time, so a previous
+            // build can start running while we sleep.
             let wait_to_build = { // Release lock before we sleep
                 let config = self.config.lock().unwrap();
                 config.wait_to_build
             };
-            
             thread::sleep(Duration::from_millis(wait_to_build));
+
+            let &mut (_, ref mut low) = &mut *self.queued_start_times.lock().unwrap();
+            if low.is_none() || low.unwrap() < start_time {
+                *low = Some(start_time);
+            }
+        } else {
+            let &mut (ref mut high, _) = &mut *self.queued_start_times.lock().unwrap();
+            if high.is_none() || high.unwrap() < start_time {
+                *high = Some(start_time);
+            }
         }
 
-        if self.running.load(Ordering::SeqCst) {
-            // Blocks
-            // println!("blocked on build");
-            let signal = rx.recv().unwrap_or(Signal::Build);
-            if signal == Signal::Skip {
-                return BuildResult::Squashed;
+        // Blocks while another thread is building.
+        let &mut (ref mut prev_high_start_time, ref mut prev_low_start_time) = &mut *self.prev_start_times.lock().unwrap();
```

Comment:
It seems unintuitive that `prev_start_times` is the main 'build rights' lock; also I believe it's possible that we might have a very high amount of extra spawned threads waiting for the finished build? E.g. when first opening the project, waiting for the analysis to finish while working on the project, where every keystroke spawns a new thread?